### ISSUE TITLE
Timeouts while calculating routing

### ISSUE THREAD

**Issue by rinigus** [otsaloma/poor-maps#23]
Hi! 

I am working on getting routing provided by OSM Scout Server. While calculations of routing within Estonia are fast (few seconds), calculations for larger distances (Aarhus-Stockholm, for example) could take 25-40 seconds. During this time, Poor Maps timeouts the connection, unfortunately.

Would it be reasonable to remove timeouts for localhost connections? Or set them rather large, like 10 minutes? I expect that timeout handling is different from the inability to establish connection (no local server running), right? If timeouts are handled differently then it should be no problem to make such change in the handling since local internet connection is available always.

Best wishes,

rinigus

**Comment by otsaloma** [comment]
25-40 seconds sounds almost like something is wrong. Have you checked if that performance is normal for libosmscout?

> I expect that timeout handling is different from the inability to establish connection (no local server running), right?

I'd expect so as well, and based on a quick test, I get an immediate `ConnectionRefusedError` if I try requesting a locahost URL without a server listening.

Yes, I could easily bump the timeout for localhost connections, but regarding a very long timeout, in general, I see two possible problems.
1. If the user chooses to not wait for the routing call to finish and swipes back a page, would he be able make a new request before the previous has finished? On the Python side, the default HTTP pool for `poor.http.request_url` and `poor.http.request_json` is limited to one connection per host. I could bump that as well to e.g. 10 for localhost URLs if it helps. On the QML side, I don't know what happens, do try and let me know!
2. If there's a routing request still in progress and the user tries to quit Poor Maps, will it hang? Poor does try to close all HTTP connections when quitting, but I don't know if it works. Do try that as well!

For testing, it should be easy for you to change these two lines:
- Timeout: https://github.com/otsaloma/poor-maps/blob/master/poor/http.py#L86
- Pool size: https://github.com/otsaloma/poor-maps/blob/master/poor/http.py#L117

**Comment by rinigus** [comment]
Re time required for calculations: Well, I think its expected. Its about 1000 km route and, as far as I remember, on OsmAnd that would take quite a few minutes to calculate. On maps.me it seemed to be much faster (order of seconds), as far as I remember. I guess different engines do have a different performance and maybe its possible to tune libosmscout. Within Sweden, I was getting routes in few seconds...

Timeout problems: yes, it turns out that the hanging connections don't allow to stop Poor Maps. User does not see it, but if you start Poor Maps from command line it will exit only after the route is calculated. 

Adding new request while route is calculated would not help much - database access is allowed by one thread only and user would have to wait till the server finishes the previous request. The parallel processing is supported for tiles, but route, search and guide requests all block the working thread. To reset, you would have to stop/start the server, unfortunately.

I think that blocking closing of the connections is a bug on my side, not Poor Maps. To tackle it as well as better treatment of multiple connections (putting them in a queue and not dropping them immediately if the pool is exhausted), I will either have to change code in underlying HTTP server library (QtWebApp) or select another library. This will not change anything for clients, just would make interaction with the server better. This is issue https://github.com/rinigus/osmscout-server/issues/9

As for long processing, I plan to give some feedback on current state of the server. That way, user will at least know that its busy calculating something. This is issue https://github.com/rinigus/osmscout-server/issues/19 

However, I think having an option for longer timeouts in Poor Maps would be needed as well. Otherwise I would be hitting them for long routes.

**Comment by otsaloma** [comment]
> Timeout problems: yes, it turns out that the hanging connections don't allow to stop Poor Maps. User does not see it, but if you start Poor Maps from command line it will exit only after the route is calculated.

It seems my code waits for connections to become idle before closing them. We can increase the timeout, but I'll want to fix the quit issue along with it.

**Comment by rinigus** [comment]
I see. Well, I guess it will be the best to fix it on the both sides. I'll release the current version after fixing some small problems to let you test, if you wish, in your setup as well. I plan to work on server's HTTP internals with the next cycle, so it should get better soon.

**Comment by rinigus** [comment]
On Fri, Oct 21, 2016 at 11:26 PM, Osmo Salomaa notifications@github.com
wrote:

> 25-40 seconds sounds almost like something is wrong. Have you checked if
> that performance is normal for libosmscout?
> 
> Well, I think its expected. Its about 1000 km route and, as far as I
> remember, on OsmAnd that would take quite a few minutes to calculate. On
> maps.me it seemed to be much faster (order of seconds), as far as I
> remember. I guess different engines do have a different performance and
> maybe its possible to tune libosmscout. Within Sweden, I was getting routes
> in few seconds...

I'll test the corner cases you mention below tomorrow. Thank you for
sending the timeout definition.

Best wishes,

rinigus

**Comment by otsaloma** [comment]
I actually just made a fix: aabe4e9be5878002e94eb1493e9b6c0280bd0a17. I expect the longer timeout should now work, but the connection termination when quitting needs testing.

**Comment by rinigus** [comment]
Thank you! I am planning to work on the web engine now to get at least connection queues right. Let's see if it will help with the connection termination as well.

**Comment by otsaloma** [comment]
Oops, it was the wrong way around: 48360361f46905a3b6a0b45a81c59195b636cdba

**Comment by rinigus** [comment]
I am testing a new backend library for providing HTTP server component. It would take some time still to learn it and change in the server, but it seems to me that we'll be able to get rid of limiting connections from clients (threads will be managed on server and all connections queued for processing, waiting for free threads if it needs to be) and I should be able to handle disconnects properly as well (without blocking for too long).

**Comment by otsaloma** [comment]
OK. File a new issue about the HTTP thread count when you have the new server.

**Comment by rinigus** [comment]
NP. Although, your code will work fine as it is. Its just you may want to simplify it if you wish :). I'll let you know when a new server is ready.

**Comment by otsaloma** [comment]
Timeout and also connection closing when quitting seemed to work fine in testing. Closing.
